### PR TITLE
feat(plugins): add plugin system

### PR DESCRIPTION
**Pull request by Niktux**
Plugin can provide configuration, twig views directories and service providers

### CODE REVIEW

**Comment by Niktux** [comment]
DO NOT MERGE
wip (missing log, unit tests, doc, ...)

File: tests/PluginManagerTest.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        ]);
+
+        $this->stackedConfiguration = new Stacked();
+        $this->stackedConfiguration->overrideBy($this->configuration);
+
+        $this->viewManager = new class implements ViewManager {
+            public $count = 0;
+            public function addPath($paths, bool $prioritary = true): void
+            {
+                $this->count += count($paths);
```
Comment (lebris) [inline tests/PluginManagerTest.php]: inutilisé

File: tests/PluginManagerTest.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        };
+
+        $this->serviceContainer = new class implements ServiceContainer {
+            public $count = 0;
+            public function register(ServiceProviderInterface $provider, array $values = array())
+            {
+                $this->count++;
```
Comment (lebris) [inline tests/PluginManagerTest.php]: inutilisé

File: tests/PluginManagerTest.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+class Unicorn extends AbstractPlugin
+{
+    public function getName(): string { return "Unicorn"; }
+
+    public function getConfiguration(): ?Configuration
+    {
+        return new Memory(['poop' => 'rainbow', 'shared' => 'unicorn', 'global' => -2]);
+    }
+}
+
```
Comment (lebris) [inline tests/PluginManagerTest.php]: \+ une classe sans `getViewDirectories` ?

File: src/Application.php [hunk 4]
```diff
@@ -56,5 +58,20 @@ protected function initializeServices(): void
     {
     }
 
+    private function initializePlugins(): void
+    {
+        $this['plugin.manager'] = function() {
+            $viewManager = null;
+            if(isset($this['view.manager']))
```
Comment (Niktux) [inline src/Application.php]: > This method is executed when using isset() or empty() on objects implementing ArrayAccess.

http://php.net/manual/en/arrayaccess.offsetexists.php

File: tests/PluginManagerTest.php (hunk 3 above)
Comment (Niktux) [inline tests/PluginManagerTest.php]: extends AbstractPlugin

File: tests/PluginManagerTest.php (hunk 1 above)
Comment (Niktux) [inline tests/PluginManagerTest.php]: WIP ;-)

File: tests/PluginManagerTest.php (hunk 2 above)
Comment (Niktux) [inline tests/PluginManagerTest.php]: WIP ;-)

**Comment by Niktux** [comment]
Missing logs

**Pull request merged**